pkg/ruler/remotequerier.go

```diff
@@ -0,0 +1,343 @@
+		},
+	}
+
+	for _, mdw := range q.middlewares {
+		if err := mdw(ctx, &req); err != nil {
+			return model.ValNone, nil, err
+		}
+	}
+
+	resp, err := q.client.Handle(ctx, &req)
+	if err != nil {
+		level.Warn(log).Log("msg", "failed to remotely evaluate query expression", "err", err, "qs", query, "tm", ts)
+		return model.ValNone, nil, err
+	}
+	level.Debug(log).Log("msg", "query expression successfully evaluated", "qs", query, "tm", ts)
+
+	if resp != nil && resp.Code/100 != 2 {
+		return model.ValNone, nil, fmt.Errorf("unexpected response status code %d: %s", resp.Code, string(resp.Body))
+	}
+
+	var apiResp struct {
+		Status    string          `json:"status"`
+		Data      json.RawMessage `json:"data"`
+		ErrorType string          `json:"errorType"`
+		Error     string          `json:"error"`
+	}
+	if err := json.NewDecoder(bytes.NewReader(resp.Body)).Decode(&apiResp); err != nil {
+		return model.ValNone, nil, err
+	}
+	if apiResp.Status == statusError {
+		return model.ValNone, nil, fmt.Errorf("query response error: %s", apiResp.Error)
+	}
+	v := struct {
+		Type   model.ValueType `json:"resultType"`
+		Result json.RawMessage `json:"result"`
+	}{}
+
+	if err := json.Unmarshal(apiResp.Data, &v); err != nil {
+		return model.ValNone, nil, err
+	}
+	return v.Type, v.Result, nil
+}
+
+// QueryFunc returns a rules.QueryFunc derived from a RemoteQuerier instance.
+func (q *RemoteQuerier) QueryFunc() rules.QueryFunc {
+	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
+		valTyp, res, err := q.Query(ctx, qs, t)
+		if err != nil {
+			return nil, err
+		}
+		return decodeQueryResponse(valTyp, res)
+	}
+}
+
+func decodeQueryResponse(valTyp model.ValueType, result json.RawMessage) (promql.Vector, error) {
+	switch valTyp {
+	case model.ValScalar:
+		var sv model.Scalar
+		if err := json.Unmarshal(result, &sv); err != nil {
+			return nil, err
+		}
+		return scalarToPromQLVector(&sv), nil
+
+	case model.ValVector:
+		var vv model.Vector
+		if err := json.Unmarshal(result, &vv); err != nil {
+			return nil, err
+		}
+		return vectorToPromQLVector(vv), nil
+
+	default:
+		return nil, fmt.Errorf("rule result is not a vector or scalar: %q", valTyp)
+	}
+}
+
+func vectorToPromQLVector(vec prommodel.Vector) promql.Vector {
+	var retVal promql.Vector
+	for _, p := range vec {
+		var sm promql.Sample
+
+		sm.V = float64(p.Value)
+		sm.T = int64(p.Timestamp)
+
+		var lbl labels.Labels
+		for ln, lv := range p.Metric {
+			lbl = append(lbl, labels.Label{Name: string(ln), Value: string(lv)})
+		}
+		sm.Metric = lbl
+
+		retVal = append(retVal, sm)
+	}
+	return retVal
+}
+
+func scalarToPromQLVector(sc *prommodel.Scalar) promql.Vector {
+	return promql.Vector{promql.Sample{
+		Point: promql.Point{
+			V: float64(sc.Value),
+			T: int64(sc.Timestamp),
+		},
+		Metric: labels.Labels{},
+	}}
+}
+
+func formatQueryTime(t time.Time) string {
+	return strconv.FormatFloat(float64(t.Unix())+float64(t.Nanosecond())/1e9, 'f', -1, 64)
+}
```

Comment:
TIL: prometheus API accepts `RFC3339Nano` format.

Reply:
I took as reference Prometheus API [documentation](https://prometheus.io/docs/prometheus/latest/querying/api/) in which they only mention `RFC3339`, and in case you need sub-second precision they suggest to use unix timestamp.

Just checked that internally we [support](https://github.com/grafana/mimir/blob/main/pkg/util/time.go#L46)  `RFC3339Nano` , so I'll go with it. 👍

Reply:
Yeah, I think that:
> with optional decimal places for sub-second precision
Is meant for both RFC3339 and Unix timestamp in that sentence.